sklearn/pipeline.py

```diff
@@ -220,20 +236,44 @@ def _final_estimator(self):
 
     def _fit(self, X, y=None, **fit_params):
         self._validate_steps()
+        # Setup the memory
+        memory = self.memory
+        if memory is None:
+            memory = Memory(cachedir=None, verbose=0)
+        elif isinstance(memory, six.string_types):
+            memory = Memory(cachedir=memory, verbose=0)
+        elif not isinstance(memory, Memory):
+            raise ValueError('memory is either `str` or a `joblib.Memory`'
+                             ' instance, got: %s' % memory)
+        # Decorate _fit_transform_one to be able to use it
+        # with cache
+        fit_transform_one = memory.cache(_fit_transform_one)
+
         fit_params_steps = dict((name, {}) for name, step in self.steps
                                 if step is not None)
         for pname, pval in six.iteritems(fit_params):
             step, param = pname.split('__', 1)
             fit_params_steps[step][param] = pval
         Xt = X
-        for name, transform in self.steps[:-1]:
-            if transform is None:
+        for step_idx, (name, transformer) in enumerate(self.steps[:-1]):
+            if transformer is None:
                 pass
-            elif hasattr(transform, "fit_transform"):
-                Xt = transform.fit_transform(Xt, y, **fit_params_steps[name])
             else:
-                Xt = transform.fit(Xt, y, **fit_params_steps[name]) \
-                              .transform(Xt)
+                if memory.cachedir is None:
+                    # we do not clone when caching is disabled to preserve
+                    # backward compatibility and the possibility to use warm
+                    # started models in that case
+                    cloned_transformer = transformer
```

Comment:
+1 for changing the comment and just stating that this is to preserve backward compat